### PR TITLE
Implement writeable wiki endpoints.

### PR DESCRIPTION
**Pull request by pattisdr**
Creating a temporary PR to leave comments - can close later.

* Create wiki pages.
* Delete wiki pages.
* Rename wiki pages.
* Edit wiki page content (create new wiki version).

### CODE REVIEW

File: api/wikis/serializers.py [hunk 1]
```diff
@@ -174,5 +228,16 @@ def get_wiki_content(self, obj):
     def get_absolute_url(self, obj):
         return obj.get_absolute_url()
 
+    def create(self, validated_data):
+        wiki_page = validated_data.get('wiki_page')
+        auth = Auth(validated_data.get('user'))
+        content = validated_data.get('content')
+
+        try:
+            _, new_version = wiki_page.node.update_node_wiki(name=wiki_page.page_name, content=content, auth=auth)
```
Comment (pattisdr) [inline api/wikis/serializers.py]: Since the content field is not required in the serializer, if no content is in the request body, you'll get a `IntegrityError: null value in column "content" violates not-null constraint`

File: framework/auth/oauth_scopes.py [hunk 2]
```diff
@@ -170,17 +171,17 @@ class ComposedScopes(object):
...
                      CoreScopes.NODE_CITATIONS_WRITE, CoreScopes.NODE_COMMENTS_WRITE, CoreScopes.NODE_FORKS_WRITE,
-                     CoreScopes.NODE_PREPRINTS_WRITE)
+                     CoreScopes.NODE_PREPRINTS_WRITE, CoreScopes.WIKI_BASE_WRITE)
 
     # Organizer Collections collection
     # Using Organizer Collections and the node links they collect. Reads Node Metadata.
     ORGANIZER_READ = (CoreScopes.ORGANIZER_COLLECTIONS_BASE_READ,) + NODE_METADATA_READ
     ORGANIZER_WRITE = ORGANIZER_READ + (CoreScopes.ORGANIZER_COLLECTIONS_BASE_WRITE, CoreScopes.NODE_LINKS_WRITE)
 
     # Privileges relating to editing content uploaded under that node # TODO: Add wiki etc when implemented
```
Comment (pattisdr) [inline framework/auth/oauth_scopes.py]: Can the TODO be removed?

File: api/wikis/serializers.py [hunk 3]
```diff
@@ -99,6 +111,42 @@ class NodeWikiSerializer(WikiSerializer):
...
+            raise Conflict(err.message)
+        return instance
+
+    def create(self, validated_data):
+        auth = Auth(validated_data.get('user'))
+        node = validated_data.get('node')
+        name = validated_data.get('page_name')
+
+        try:
+            node.wikis.get(page_name__iexact=name, deleted__isnull=True)
```
Comment (pattisdr) [inline api/wikis/serializers.py]: Can you use the node.get_wiki_page method instead of duplicating that query here?

File: api/nodes/views.py [hunk 4]
```diff
@@ -2889,6 +2889,9 @@ def get_default_queryset(self):
     def get_queryset(self):
         return self.get_queryset_from_request()
 
+    def perform_create(self, serializer):
+        return serializer.save(node=self.get_node(), user=self.request.user)
```
Comment (pattisdr) [inline api/nodes/views.py]: What is the purpose of passing in the user when you can just do `Auth(self.context['request'].user)` later?

File: api/wikis/serializers.py [hunk 5]
```diff
@@ -174,5 +228,16 @@ def get_wiki_content(self, obj):
...
+        content = validated_data.get('content')
+
+        try:
+            _, new_version = wiki_page.node.update_node_wiki(name=wiki_page.page_name, content=content, auth=auth)
+        except NodeStateError:
+            raise MethodNotAllowed(method=self.context['request'].method)
+        return new_version
+
     class Meta:
         type_ = 'wiki-versions'
```
Comment (pattisdr) [inline api/wikis/serializers.py]: We're going to need a TypeField() on this WikiVersion serializer to enforce that type is wiki-versions.

File: api/nodes/views.py [hunk 6]
```diff
@@ -2820,8 +2820,8 @@ def create(self, *args, **kwargs):
         return ret
 
 
-class NodeWikiList(JSONAPIBaseView, generics.ListAPIView, NodeMixin, ListFilterMixin):
-    """List of wiki pages on a node. *Read only*.
+class NodeWikiList(JSONAPIBaseView, generics.ListCreateAPIView, NodeMixin, ListFilterMixin):
+    """List of wiki pages on a node.
```
Comment (pattisdr) [inline api/nodes/views.py]: Have you added docs? I've started some wiki docs here that could be added onto: https://github.com/CenterForOpenScience/developer.osf.io/pull/7

File: api/wikis/serializers.py [hunk 7]
```diff
@@ -99,6 +111,42 @@ class NodeWikiSerializer(WikiSerializer):
...
+            node.wikis.get(page_name__iexact=name, deleted__isnull=True)
+        except WikiPage.DoesNotExist:
+            pass
+        else:
+            raise Conflict("A wiki page with the name '{}' already exists.".format(name))
+
+        try:
+            wiki_page, _ = node.update_node_wiki(name=name, content='', auth=auth)
+        except (
+            NameEmptyError, NameInvalidError,
```
Comment (pattisdr) [inline api/wikis/serializers.py]: Probably not possible to get a NameEmptyError since you made `name` a required serializer field?

File: api/wikis/serializers.py [hunk 8]
```diff
@@ -99,6 +111,42 @@ class NodeWikiSerializer(WikiSerializer):
...
+            NameMaximumLengthError, django_exceptions.ValidationError
```
Comment (pattisdr) [inline api/wikis/serializers.py]: When is this django_exceptions.ValidationError thrown?

File: api/wikis/serializers.py [hunk 9]
```diff
@@ -99,6 +111,42 @@ class NodeWikiSerializer(WikiSerializer):
...
+
+        try:
+            node.wikis.get(page_name__iexact=name, deleted__isnull=True)
+        except WikiPage.DoesNotExist:
+            pass
+        else:
+            raise Conflict("A wiki page with the name '{}' already exists.".format(name))
+
+        try:
+            wiki_page, _ = node.update_node_wiki(name=name, content='', auth=auth)
```
Comment (pattisdr) [inline api/wikis/serializers.py]: I am wondering if we should allow "content" to be in the request body here on wiki creation only.  I know this endpoint is for creating a wiki page, but since a wiki version 1 is created in the background anyway, this means that everyone's wiki version one will be blank if created via the API.  This way adding content to a wiki the first time doesn't necessarily have to be a two-step process. What do you think?

File: api/wikis/serializers.py [hunk 10]
```diff
@@ -99,6 +111,42 @@ class NodeWikiSerializer(WikiSerializer):
         related_view_kwargs={'wiki_id': '<_id>'},
     )
 
+    def update(self, instance, validated_data):
+        auth = Auth(self.context['request'].user)
+        new_page_name = validated_data.get('page_name')
+        try:
+            instance = instance.node.rename_node_wiki(instance.page_name, new_page_name, auth)
+        except (
+            NameEmptyError, NameInvalidError, NodeStateError,
```
Comment (pattisdr) [inline api/wikis/serializers.py]: Similar here, don't expect NameEmptyError to ever be thrown.

File: osf/models/node.py [hunk 11]
```diff
@@ -2813,8 +2815,15 @@ def rename_node_wiki(self, name, new_name, auth):
...
-        page = self.get_wiki_page(name)
+        return page
+
+    def delete_node_wiki(self, name_or_page, auth):
+        WikiPage = apps.get_model('addons_wiki.WikiPage')
+        page = name_or_page
+        if not isinstance(name_or_page, WikiPage):
+            page = self.get_wiki_page(name_or_page)
+        if page.page_name.lower() == 'home':
+            raise ValidationError('The home wiki page cannot be deleted.')
```
Comment (pattisdr) [inline osf/models/node.py:2848]: 👍

File: api/wikis/views.py [hunk 12]
```diff
@@ -131,6 +141,17 @@ def get_serializer_class(self):
...
         return self.get_wiki()
 
+    # overrides RetrieveUpdateDestroyAPIView
+    def perform_destroy(self, instance):
+        auth = get_user_auth(self.request)
+        wiki_page = self.get_object()
+        try:
+            wiki_page.node.delete_node_wiki(wiki_page, auth)
+        except NodeStateError:
+            raise MethodNotAllowed(method=self.request.method)
```
Comment (pattisdr) [inline api/wikis/views.py]: Just a question - it took me awhile to figure out what would be throwing a NodeStateError until I saw you'd added it to the Registration model.  I am just wondering if this is the best way to do this.  How do we handle other objects that can belong to either a node or a registration in the API?

Was thinking about how we inherit from a generic ListCreateView, for example, to restrict what methods we could use. Maybe we could create something similar for these wiki views to inherit from to restrict the http methods if the wiki's node is a reg. Seems like we could consolidate this NodeStateError checking.

Feel free to push back.

**Review by pattisdr — COMMENTED** [review]
Excellent start @caseyrollins, thanks for all the great test coverage.  You can tell this isn't the first time you've implemented some major API endpoints!

Apparently I cannot request changes on my own PR :)

File: api/nodes/views.py (hunk 6 above)
Comment (caseyrollins) [reply api/nodes/views.py]: Not yet, it's in my TODOs. I'll merge your PR first so I can work off of that.

File: api/wikis/views.py (hunk 12 above)
Comment (caseyrollins) [reply api/wikis/views.py]: @pattisdr, this is an excellent note. I went back and forth on this as well, but I think it's most appropriate for validation to happen at the model level, not in the view (Chris Wisecarver beat the "skinny views, fat models" idea into me). Another reason I think it's preferable to have this validation at the model level is so that if someone attempts to call `delete_node_wiki` (or rename or edit) outside of the API (e.g. in a script), this error will be raised. However, I definitely agree that the `NodeStateError` is not easy to track down and I'm not sure how to improve that. 

Let me know if you have other thoughts on this. It's certainly an opinionated decision where I didn't consider all possibilities (such as your mixin suggestion) 

From a (short) FD discussion with Steve:
```
Casey
Question of the day:
Wiki pages on registrations can’t be deleted. To prevent this from happening in the API, do we:
a) check is_registration in AbstractNode.delete_node_wiki and raise an exception that is handled by the API
b) override delete_node_wiki on the Registration model to raise an exception that is handled by the API
c) check wiki_page.node.is_registration in the WikiDetail API view and raise an exception

Casey
I would vote b

Casey
But we don’t seem to be doing that anywhere?

Steve
I second b
```

File: api/wikis/serializers.py (hunk 10 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: Good catch.

File: api/wikis/serializers.py (hunk 3 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: Yep, good catch! I must've just forgotten about it here.

File: api/wikis/serializers.py (hunk 8 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: I dug into this for a little while and still have no idea. Maybe I got confused and was thinking of the error that can be raised from deleting the home wiki page? I'll remove and see if any tests fail.

File: api/nodes/views.py (hunk 4 above)
Comment (caseyrollins) [reply api/nodes/views.py]: None at all; who knows why I did that.

File: api/wikis/serializers.py (hunk 1 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: Do you agree that content should remain not required and we should just add the following check?
```
if not content:
    content = ''
```

Comment (pattisdr) [reply api/wikis/serializers.py]: Yes, that is fine!

Comment (caseyrollins) [reply api/wikis/serializers.py]: Either way, I'll add the appropriate test for this case.

File: api/wikis/views.py (hunk 12 above)
Comment (pattisdr) [reply api/wikis/views.py]: @caseyrollins, I think whether or not we override `delete_node_wiki` and others on the Registration model is a separate issue. Whichever way you handle things API side, I think this was a good idea to further safeguard registrations.

The main reason I commented here, though, was after noticing the numerous NodeStateError checks across your views that occur well after the point in which we know the request will fail. For example, we've already validated the request data, which is unnecessary work if this is a disallowed request. As soon as the wiki is fetched, we have its node or registration, and therefore know if the request_method is acceptable.  

Similarly, there is no way to delete a registration through the API, we use a RetrieveUpdateAPIView on RegistrationDetail.  As soon as we send a Delete request to that endpoint it fails.  We don't have to wait to call the delete method on the registration to decide to throw a 405.

Comment (caseyrollins) [reply api/wikis/views.py]: Ah, I see what you mean. Not all of my `NodeStateError` checks are necessary (they'll never be hit, as you're saying). 

Here are the existing checks:
* [`NodeWikiSerializer.update()`](https://github.com/pattisdr/osf.io/pull/10/files#diff-dfa23dfe3da2849868aa007a7435fe5bR120): Definitely not needed as the request to update a registration wiki page would fail first [here](https://github.com/pattisdr/osf.io/pull/10/files#diff-dfa23dfe3da2849868aa007a7435fe5bR165). 
* [`WikiVersionSerializer.create()`](https://github.com/pattisdr/osf.io/pull/10/files#diff-dfa23dfe3da2849868aa007a7435fe5bR165): Arguably useful since both nodes and registrations share this serializer.
* [`WikiDetail.perform_destroy()`](https://github.com/pattisdr/osf.io/pull/10/files#diff-15e45414bd0cfad4288438347ef25273R150): Arguably useful since both node and registration wiki pages share this view.

However, I just now realized that the last two could be checked in [`WikiMixin.get_wiki()`](https://github.com/CenterForOpenScience/osf.io/blob/develop/api/wikis/views.py#L28) instead which will save us the wasted effort of validating request data, as you say. Let's go with that?

Comment (pattisdr) [reply api/wikis/views.py]: That works great @caseyrollins , thanks

File: api/wikis/serializers.py (hunk 1 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: Ended up adding `allow_blank=False` to the `content` fields in the wiki and wiki version serializers.

File: api/wikis/serializers.py (hunk 9 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: I agree! I've added a write only content field that is only set on creation.

**Comment by caseyrollins** [comment]
@pattisdr, thank you for the thorough review! Ready for another pass!

File: api/wikis/views.py [hunk 13]
```diff
@@ -36,6 +45,9 @@ def get_wiki(self, check_permissions=True):
         if wiki.deleted:
             raise Gone
 
+        if wiki.node.is_registration and self.request.method not in drf_permissions.SAFE_METHODS:
+            raise MethodNotAllowed(method=self.request.method)
```
Comment (pattisdr) [inline api/wikis/views.py:49]: 💯

File: api/wikis/serializers.py [hunk 14]
```diff
@@ -99,6 +108,41 @@ class NodeWikiSerializer(WikiSerializer):
...
+        auth = Auth(self.context['request'].user)
+        node = validated_data.get('node')
+        name = validated_data.get('page_name')
+        content = validated_data.get('content', '')
+
+        if node.get_wiki_page(name=name):
+            raise Conflict("A wiki page with the name '{}' already exists.".format(name))
+
+        try:
+            wiki_page, _ = node.update_node_wiki(name=name, content=content, auth=auth)
```
Comment (pattisdr) [inline api/wikis/serializers.py:136]: I should know this, but I remember there being some bleach stuff in the wiki code, do we need to do anything to the content before saving it to the db?

File: api/wikis/serializers.py [hunk 15]
```diff
@@ -114,6 +158,12 @@ class RegistrationWikiSerializer(WikiSerializer):
         filter={'target': '<_id>'}
     )
 
+    def update(self, instance, validated_data):
+        raise MethodNotAllowed(method=self.context['request'].method)
+
+    def create(self, validated_data):
+        raise MethodNotAllowed(method=self.context['request'].method)
```
Comment (pattisdr) [inline api/wikis/serializers.py]: Will these methods get called or is this caught earlier?

File: api/wikis/serializers.py [hunk 16]
```diff
@@ -99,6 +108,41 @@ class NodeWikiSerializer(WikiSerializer):
         related_view_kwargs={'wiki_id': '<_id>'},
     )
 
+    def update(self, instance, validated_data):
+        auth = Auth(self.context['request'].user)
+        new_page_name = validated_data.get('page_name')
+        validated_data.pop('content', None)
+
+        if new_page_name:
```
Comment (pattisdr) [inline api/wikis/serializers.py]: Is the check necessary?

**Review by pattisdr — COMMENTED** [review]
Just a couple questions @caseyrollins!

File: api/wikis/serializers.py (hunk 14 above)
Comment (caseyrollins) [reply api/wikis/serializers.py:136]: @pattisdr, I can see where we sanitize before rendering the content, but if we sanitize on input I can't find it. @sloria do you know if this exists?

File: api/wikis/serializers.py (hunk 15 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: I believe the `create` override is needed but please double check my reasoning:

`RegistrationWikiSerializer` is used in the `RegistrationWikiList` view, which inherits from the `NodeWikiList` view. The `NodeWikiList` view doesn't raise `MethodNotAllowed` for registrations like views with the `WikiMixin` do, therefore the error will be raised here.

The `update` override is now obsolete; will remove.

File: api/wikis/serializers.py (hunk 14 above)
Comment (sloria) [reply api/wikis/serializers.py:136]: Serializers sanitize input upon validation: https://github.com/CenterForOpenScience/osf.io/blob/86a567fcfb71c6ed03fd3611c2060203cad971e6/api/base/serializers.py#L1120

Comment (pattisdr) [reply api/wikis/serializers.py:136]: That's fine, thanks for looking, just wanted to double-check.

File: api/wikis/serializers.py (hunk 16 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: Nope! I think this was from before I added `allow_blank=False` to `name`. Will remove.

**Comment by pattisdr** [comment]
Looks good @caseyrollins

File: api/wikis/views.py [hunk 17]
```diff
@@ -131,6 +143,15 @@ def get_serializer_class(self):
     def get_object(self):
         return self.get_wiki()
 
+    # overrides RetrieveUpdateDestroyAPIView
+    def perform_destroy(self, instance):
+        auth = get_user_auth(self.request)
+        wiki_page = self.get_object()
+        try:
+            wiki_page.node.delete_node_wiki(wiki_page, auth)
```
Comment (sloria) [inline api/wikis/views.py:151]: Drive-by: Perhaps we should refactor to put the `delete...` method on `wiki_page` instead of node (see [Law of Demeter](https://www.wikiwand.com/en/Law_of_Demeter)).

File: osf/models/node.py [hunk 18]
```diff
@@ -2813,8 +2815,15 @@ def rename_node_wiki(self, name, new_name, auth):
             save=True,
         )
 
-    def delete_node_wiki(self, name, auth):
-        page = self.get_wiki_page(name)
+        return page
+
+    def delete_node_wiki(self, name_or_page, auth):
```
Comment (sloria) [inline osf/models/node.py:2842]: See previous comment: this should probably go on `WikiPage`. We want to remove as much wiki-related code from the node as possible.

File: api/wikis/serializers.py [hunk 19]
```diff
@@ -99,6 +108,40 @@ class NodeWikiSerializer(WikiSerializer):
         related_view_kwargs={'wiki_id': '<_id>'},
     )
 
+    def update(self, instance, validated_data):
+        auth = Auth(self.context['request'].user)
+        new_page_name = validated_data.get('page_name')
+        validated_data.pop('content', None)
+
+        try:
+            instance = instance.node.rename_node_wiki(instance.page_name, new_page_name, auth)
```
Comment (sloria) [inline api/wikis/serializers.py:117]: Same here: move `rename_node_wiki` to `Node`?

File: api/wikis/serializers.py [hunk 20]
```diff
@@ -99,6 +108,40 @@ class NodeWikiSerializer(WikiSerializer):
...
+        except (
+            NameInvalidError, NameMaximumLengthError,
+            PageCannotRenameError
```
Comment (sloria) [inline api/wikis/serializers.py]: Rather than putting handling all these, you could do

```python
except PageConflict as err:
    raise Conflict(...)
except WikiError...:
   raise ValidationError(...)
```

File: api/wikis/serializers.py [hunk 21]
```diff
@@ -99,6 +108,40 @@ class NodeWikiSerializer(WikiSerializer):
...
+        ) as err:
+            raise ValidationError(err.message)
```
Comment (sloria) [inline api/wikis/serializers.py]: Are you sure that `err.message` is an attribute? `Exception` objects don't have a `message` attribute unless it's explicitly set on them.

File: api/nodes/views.py [hunk 22]
```diff
@@ -2889,6 +2889,9 @@ def get_default_queryset(self):
     def get_queryset(self):
         return self.get_queryset_from_request()
 
+    def perform_create(self, serializer):
+        return serializer.save(node=self.get_node())
```
Comment (sloria) [inline api/nodes/views.py:1442]: Is `node` used in `NodeWikiSerializer#save`?

Comment (caseyrollins) [reply api/nodes/views.py:1442]: See http://www.django-rest-framework.org/topics/3.0-announcement/#changes-to-prepost-save-hooks and http://www.django-rest-framework.org/api-guide/serializers/#saving-instances. 

`perform_create` calls `serializer.save()` which passes `node` to `NodeWikiSerializer.create()`

File: api/wikis/serializers.py (hunk 20 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: Ah, this looks so much better.

**Review by sloria — COMMENTED** [review]
Pass finished. No major blockers, though you should double-check where I commented on exception handling.

I suggested moving some of the wiki functionality from Node to WikiPage, but I now see that existing wiki functionality is already on Node and you're following that. Feel free to move the code if you're feeling up to it. This can't get merged until the modeling PR goes in, so you've got a little time. I'm fine with this as is, though.

File: api/wikis/serializers.py (hunk 21 above)
Comment (caseyrollins) [reply api/wikis/serializers.py]: Yep. `NameInvalidError`, `NameMaximumLengthError`, and `PageCannotRenameError` all set message.

https://github.com/CenterForOpenScience/osf.io/blob/develop/addons/wiki/models.py#L42
https://github.com/CenterForOpenScience/osf.io/blob/develop/osf/models/node.py#L2811

Although it doesn't hurt to add a check in case the messages are removed. Is there a cleaner way than:
```
...
except WikiError as err:
    if getattr(err, 'message'):
        raise ValidationError(err.message)
    raise ValidationError
```

**Comment by caseyrollins** [comment]
While a bit out of scope, I definitely agree that those wiki functions belong on the WikiPage model, not Node. I'll see if it's simple to move since I've got the time.

**Comment by caseyrollins** [comment]
@sloria, on second thought, I think moving the wiki functions from `Node` to `WikiPage` should be deferred (I've made a [ticket]()). There are several functions that I think should be moved:
- `delete_node_wiki`

**Comment by caseyrollins** [comment]
@sloria, on second thought, I think moving the wiki functions from `Node` to `WikiPage` should be deferred (I've made a [ticket](https://openscience.atlassian.net/browse/PLAT-746) for it). Since the node-wiki helpers are closely tied to one another, I think it makes sense to move them all at once. However, since they're all widely used (especially in tests), making this change would result in a lot of changed code which I don't think is appropriate for this PR. 

`get_wiki_page` 
- 57 occurrences to update

`delete_node_wiki` 
- 14 occurrences 
- calls `get_wiki_page`

`rename_node_wiki` 
- 16 occurrences 
- calls `get_wiki_page`

`create_or_update_node_wiki`
- 5 occurrences
- calls `get_wiki_page`

`update_node_wiki` 
- 92 occurrences 
- calls `create_or_update_node_wiki`

**Comment by sloria** [comment]
Fine by me!

**Comment by pattisdr** [comment]
Closing, this was just an initial PR to get comments, actual PR was merged.
